src/xwayland/xserver.rs

```diff
@@ -290,72 +291,79 @@ fn xwayland_ready<WM: XWindowManager>(inner: &Rc<RefCell<Inner<WM>>>) {
     }
 }
 
-pub(crate) enum Void {}
-
 /// Exec XWayland with given sockets on given display
 ///
 /// If this returns, that means that something failed
-pub(crate) fn exec_xwayland(
+fn spawn_xwayland(
     display: u32,
     wayland_socket: UnixStream,
     wm_socket: UnixStream,
     listen_sockets: &[UnixStream],
-) -> NixResult<Void> {
-    // uset the CLOEXEC flag from the sockets we need to pass
-    // to xwayland
-    unset_cloexec(&wayland_socket)?;
-    unset_cloexec(&wm_socket)?;
-    for socket in listen_sockets {
-        unset_cloexec(socket)?;
-    }
-    // prepare the arguments to XWayland
-    let mut args = vec![
-        CString::new("Xwayland").unwrap(),
-        CString::new(format!(":{}", display)).unwrap(),
-        CString::new("-rootless").unwrap(),
-        CString::new("-terminate").unwrap(),
-        CString::new("-wm").unwrap(),
-        CString::new(format!("{}", wm_socket.as_raw_fd())).unwrap(),
-    ];
+) -> IOResult<ChildStdout> {
+    let mut command = Command::new("sh");
+
+    // We use output stream to communicate because FD is easier to handle than exit code.
+    command.stdout(Stdio::piped());
+
+    let mut xwayland_args = format!(":{} -rootless -terminate -wm {}", display, wm_socket.as_raw_fd());
     for socket in listen_sockets {
-        args.push(CString::new("-listen").unwrap());
-        args.push(CString::new(format!("{}", socket.as_raw_fd())).unwrap());
+        xwayland_args.push_str(&format!(" -listen {}", socket.as_raw_fd()));
     }
-    // setup the environment: clear everything except PATH and XDG_RUNTIME_DIR
-    for (key, _) in env::vars_os() {
+    // This command let sh to:
+    // * Set up signal handler for USR1
+    // * Launch Xwayland with USR1 ignored so Xwayland will signal us when it is ready (also redirect
+    //   Xwayland's STDOUT to STDERR so its output, if any, won't distract us)
+    // * Print "S" and exit if USR1 is received
+    command.arg("-c").arg(format!(
```

Comment:
Isn't `format!` allocating memory? That's not `fork`-safe (I didn't look at the full PR, only this part)

Reply:
There are no forks anymore, only spawn.

Reply:
Oh nice. That's actually quite neat.